File: notes/mcp/tools.js
import { z } from "zod";
import { NOTES_TOOL_NAMES } from "./tool-catalog.js";

const noteText = z.string().trim().min(1).max(100 * 1024);
const tags = z.array(z.string().trim().min(1).max(60)).max(20).default([]);
const noteId = z.number().int().positive();
const searchQuery = z.string().trim().min(1).max(500);
const tagFilter = z.string().trim().min(1).max(60).optional();
const sortOption = z.enum(["updated", "created"]).default("updated");

const schemas = {
  create: z.object({ text: noteText, tags }).strict(),
  list: z.object({ tag: tagFilter, sort: sortOption }).strict(),
  get: z.object({ id: noteId }).strict(),
  update: z.object({
    id: noteId,
    text: noteText.optional(),
    tags: z.array(z.string().trim().min(1).max(60)).max(20).optional(),
  }).strict(),
  delete: z.object({ id: noteId }).strict(),
  search: z.object({ query: searchQuery }).strict(),
};

export const NOTES_TOOLS = Object.freeze([
  descriptor(
    "create",
    "Create a new note with optional tags. Text supports markdown.",
    {
      text: { type: "string", description: "Note content (markdown supported, max 100 KB)." },
      tags: { type: "array", items: { type: "string" }, description: "Optional tags (max 20, each max 60 chars).", default: [] },
    },
    ["text"],
    false,
  ),
  descriptor(
    "list",
    "List all notes, optionally filtered by tag. Results are sorted by updatedAt (default) or createdAt.",
    {
      tag: { type: "string", description: "Filter notes by tag name. Omit to list all." },
      sort: { type: "string", enum: ["updated", "created"], description: "Sort order: 'updated' (default) or 'created'.", default: "updated" },
    },
    [],
  ),
  descriptor(
    "get",
    "Get a single note by its ID.",
    {
      id: { type: "integer", description: "Note ID (positive integer).", minimum: 1 },
    },
    ["id"],
  ),
  descriptor(
    "update",
    "Update a note's text and/or tags. Only provided fields are changed.",
    {
      id: { type: "integer", description: "Note ID to update.", minimum: 1 },
      text: { type: "string", description: "New note text (markdown supported, max 100 KB). Omit to keep existing." },
      tags: { type: "array", items: { type: "string" }, description: "New tags array (replaces existing). Omit to keep existing." },
    },
    ["id"],
    false,
  ),
  descriptor(
    "delete",
    "Delete a note by its ID. This is permanent and cannot be undone.",
    {
      id: { type: "integer", description: "Note ID to delete.", minimum: 1 },
    },
    ["id"],
    false,
  ),
  descriptor(
    "search",
    "Search notes by text content using a regex pattern (case-insensitive).",
    {
      query: { type: "string", description: "Regex pattern to match against note text (case-insensitive)." },
    },
    ["query"],
  ),
]);

if (NOTES_TOOLS.map((t) => t.name).join(",") !== NOTES_TOOL_NAMES.join(",")) {
  throw new Error("Notes tool descriptors are out of sync with the canonical catalog");
}

export async function callNotesTool(service, name, rawArguments = {}, { persist = true } = {}) {
  const schema = schemas[name];
  if (!schema) throw new Error(`Unknown notes tool: ${name}`);
  const input = schema.parse(rawArguments ?? {});

  switch (name) {
    case "create": {
      const note = service.create(input.text, input.tags);
      if (persist) await service.save();
      return { note, totalNotes: service.notes.length };
    }
    case "list": {
      const notes = service.list(input.tag, input.sort);
      return { notes, total: notes.length, ...(input.tag ? { tag: input.tag } : {}), sort: input.sort };
    }
    case "get": {
      const note = service.get(input.id);
      return { note };
    }
    case "update": {
      const note = service.update(input.id, {
        ...(input.text !== undefined ? { text: input.text } : {}),
        ...(input.tags !== undefined ? { tags: input.tags } : {}),
      });
      if (persist) await service.save();
      return { note };
    }
    case "delete": {
      const deleted = service.delete(input.id);
      if (persist) await service.save();
      return { deleted, totalNotes: service.notes.length };
    }
    case "search": {
      const results = service.search(input.query);
      return { results, total: results.length, query: input.query };
    }
    default:
      throw new Error(`Unknown notes tool: ${name}`);
  }
}

function descriptor(name, description, properties, required = [], readOnly = true) {
  return {
    name,
    description,
    annotations: {
      title: name,
      readOnlyHint: readOnly,
      destructiveHint: name === "delete",
      idempotentHint: readOnly || name === "update",
      openWorldHint: false,
    },
    inputSchema: {
      type: "object",
      properties,
      required,
      additionalProperties: false,
    },
  };
}

File: notes/mcp/tools.go
package main

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

// Tool names — must not start with "notes_" or equal "notes" (create rule).
const (
	toolCreate = "create"
	toolList   = "list"
	toolGet    = "get"
	toolUpdate = "update"
	toolDelete = "delete"
	toolSearch = "search"
)

func registerTools(s *server.MCPServer, svc *NoteService) {
	s.AddTool(
		mcp.NewTool(toolCreate,
			mcp.WithDescription("Create a new note with optional tags. Text supports markdown."),
			mcp.WithString("text",
				mcp.Required(),
				mcp.Description("Note content (markdown supported, max 100 KB)."),
				mcp.MaxLength(maxTextLength),
				mcp.MinLength(1),
			),
			mcp.WithArray("tags",
				mcp.Description("Optional tags (max 20, each max 60 chars)."),
				mcp.MaxItems(maxTags),
				mcp.Items(map[string]any{"type": "string"}),
			),
		),
		handleCreate(svc),
	)

	s.AddTool(
		mcp.NewTool(toolList,
			mcp.WithDescription("List all notes, optionally filtered by tag. Results are sorted by updatedAt (default) or createdAt."),
			mcp.WithString("tag",
				mcp.Description("Filter notes by tag name. Omit to list all."),
				mcp.MaxLength(maxTagLength),
			),
			mcp.WithString("sort",
				mcp.Description("Sort order: 'updated' (default) or 'created'."),
				mcp.Enum("updated", "created"),
			),
		),
		handleList(svc),
	)

	s.AddTool(
		mcp.NewTool(toolGet,
			mcp.WithDescription("Get a single note by its ID."),
			mcp.WithNumber("id",
				mcp.Required(),
				mcp.Description("Note ID (positive integer)."),
				mcp.Min(1.0),
			),
		),
		handleGet(svc),
	)

	s.AddTool(
		mcp.NewTool(toolUpdate,
			mcp.WithDescription("Update a note's text and/or tags. Only provided fields are changed."),
			mcp.WithNumber("id",
				mcp.Required(),
				mcp.Description("Note ID to update."),
				mcp.Min(1.0),
			),
			mcp.WithString("text",
				mcp.Description("New note text (markdown supported, max 100 KB). Omit to keep existing."),
				mcp.MaxLength(maxTextLength),
			),
			mcp.WithArray("tags",
				mcp.Description("New tags array (replaces existing). Omit to keep existing."),
				mcp.MaxItems(maxTags),
				mcp.Items(map[string]any{"type": "string"}),
			),
		),
		handleUpdate(svc),
	)

	s.AddTool(
		mcp.NewTool(toolDelete,
			mcp.WithDescription("Delete a note by its ID. This is permanent and cannot be undone."),
			mcp.WithNumber("id",
				mcp.Required(),
				mcp.Description("Note ID to delete."),
				mcp.Min(1.0),
			),
		),
		handleDelete(svc),
	)

	s.AddTool(
		mcp.NewTool(toolSearch,
			mcp.WithDescription("Search notes by text content using a regex pattern (case-insensitive)."),
			mcp.WithString("query",
				mcp.Required(),
				mcp.Description("Regex pattern to match against note text (case-insensitive)."),
				mcp.MaxLength(500),
			),
		),
		handleSearch(svc),
	)
}

func handleCreate(svc *NoteService) server.ToolHandlerFunc {
	return func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		args := req.GetArguments()
		text, _ := args["text"].(string)
		tags := toStringSlice(args["tags"])

		note, err := svc.Create(text, tags)
		if err != nil {
			return errorResult(err), nil
		}
		if err := svc.Save(); err != nil {
			return errorResult(err), nil
		}
		return jsonResult(map[string]any{
			"note":       note,
			"totalNotes": svc.Total(),
		})
	}
}

func handleList(svc *NoteService) server.ToolHandlerFunc {
	return func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		args := req.GetArguments()
		tag, _ := args["tag"].(string)
		sortKey, _ := args["sort"].(string)
		if sortKey == "" {
			sortKey = "updated"
		}

		notes := svc.List(tag, sortKey)
		result := map[string]any{
			"notes": notes,
			"total": len(notes),
			"sort":  sortKey,
		}
		if tag != "" {
			result["tag"] = tag
		}
		return jsonResult(result)
	}
}

func handleGet(svc *NoteService) server.ToolHandlerFunc {
	return func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		id, ok := toInt(req.GetArguments()["id"])
		if !ok {
			return errorResult(fmt.Errorf("id must be a positive integer")), nil
		}
		note, err := svc.Get(id)
		if err != nil {
			return errorResult(err), nil
		}
		return jsonResult(map[string]any{"note": note})
	}
}

func handleUpdate(svc *NoteService) server.ToolHandlerFunc {
	return func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		args := req.GetArguments()
		id, ok := toInt(args["id"])
		if !ok {
			return errorResult(fmt.Errorf("id must be a positive integer")), nil
		}

		var updates NoteUpdate
		if text, exists := args["text"]; exists {
			if s, ok := text.(string); ok {
				updates.Text = &s
			}
		}
		if tags, exists := args["tags"]; exists {
			slice := toStringSlice(tags)
			updates.Tags = &slice
		}

		note, err := svc.Update(id, updates)
		if err != nil {
			return errorResult(err), nil
		}
		if err := svc.Save(); err != nil {
			return errorResult(err), nil
		}
		return jsonResult(map[string]any{"note": note})
	}
}

func handleDelete(svc *NoteService) server.ToolHandlerFunc {
	return func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		id, ok := toInt(req.GetArguments()["id"])
		if !ok {
			return errorResult(fmt.Errorf("id must be a positive integer")), nil
		}
		deleted, err := svc.Delete(id)
		if err != nil {
			return errorResult(err), nil
		}
		if err := svc.Save(); err != nil {
			return errorResult(err), nil
		}
		return jsonResult(map[string]any{
			"deleted":    deleted,
			"totalNotes": svc.Total(),
		})
	}
}

func handleSearch(svc *NoteService) server.ToolHandlerFunc {
	return func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		query, _ := req.GetArguments()["query"].(string)
		results := svc.Search(query)
		return jsonResult(map[string]any{
			"results": results,
			"total":   len(results),
			"query":   query,
		})
	}
}

// --- helpers ---

func toStringSlice(v any) []string {
	if v == nil {
		return nil
	}
	arr, ok := v.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(arr))
	for _, item := range arr {
		if s, ok := item.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

func toInt(v any) (int, bool) {
	switch n := v.(type) {
	case float64:
		return int(n), true
	case int:
		return n, true
	case int64:
		return int(n), true
	default:
		return 0, false
	}
}

func jsonResult(data any) (*mcp.CallToolResult, error) {
	raw, err := json.Marshal(data)
	if err != nil {
		return nil, fmt.Errorf("marshal result: %w", err)
	}
	return &mcp.CallToolResult{
		Content: []mcp.Content{
			mcp.NewTextContent(string(raw)),
		},
		StructuredContent: data,
	}, nil
}

func errorResult(err error) *mcp.CallToolResult {
	return &mcp.CallToolResult{
		IsError: true,
		Content: []mcp.Content{
			mcp.NewTextContent(safeError(err)),
		},
	}
}

func safeError(err error) string {
	msg := err.Error()
	if len(msg) > 1000 {
		msg = msg[:1000]
	}
	return msg
}
